### PR TITLE
 Stricten, test, and bugfix http-server

### PR DESCRIPTION
**Pull request by diehuxx**
## All Changes
1. Turn on typescript `strict` mode for the `http-server` package. Fix any bugs that `strict` mode reveals.
2. Add many many tests which cover the request handlers completely, except for `callback` edge cases, which will be reworked in a follow up PR.
3. Refactor how `TbdexHttpServer` calls the `callbacks` because the previous way actually did not work.
4. Remove `fakes.ts`. Flesh out the "fake" implementations of `ExchangesApi` and `OfferingsApi` as fully usable `InMemory` implementations. See `in-memory-exchanges-api.ts` and `in-memory-offerings-api.ts`. These do not yet have 100% test coverage simply because this PR already adds ~1200 lines of tests; I'll add more tests in a follow up PR.

## Changes to external API
Add fully fleshed out `InMemoryExchangesApi` and `InMemoryOfferingsApi`. These are used as defaults in `TbdexHttpServer`, but may be worth documenting in their own right.

## Items TODO in follow up PRs
1. Reach 100% test coverage of `in-memory-exchanges-api.ts` and `in-memory-offerings-api.ts`.
2. Solidify the behavior of callback functions in the request handlers. Specifically, how callbacks can/cannot alter the HTTP response, how to handle if a callback throws an error, and what exactly should be passed to callbacks.

### CODE REVIEW

File: packages/http-server/src/http-server.ts [hunk 1]
```diff
@@ -139,25 +140,42 @@
...
+        exchangesApi,
+      })
+    )
+
+    this.api.get('/exchanges', (req: Request, res: Response) =>
+      getExchanges(req, res, {
+        callback: this.callbacks.exchanges,
+        exchangesApi,
+        pfiDid,
+      })
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/http-server.ts:170]: I'm not sure if it's appropriate for us to set up rate limiting at the `tbdex-js` level. Regardless, that is outside the scope of this PR and can be addressed in a follow up.

File: packages/http-server/src/http-server.ts [hunk 2]
```diff
@@ -139,25 +140,42 @@
...
-
-    this.api.get('/offerings', getOfferings({
-      callback: this.callbacks['offerings'], offeringsApi
-    }))
+    this.api.post('/exchanges/:exchangeId/rfq', (req: Request, res: Response) =>
+      createExchange(req, res, {
+        callback: this.callbacks['rfq'],
+        offeringsApi,
+        exchangesApi,
+      })
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/http-server.ts:148]: I'm not sure if it's appropriate for us to set up rate limiting at the `tbdex-js` level. Regardless, that is outside the scope of this PR and can be addressed in a follow up.

**Comment by diehuxx** [comment]
I'm deliberately not adding a changeset to this PR since I don't want to release a new version at this point. This PR adds a lot of tests and cleanup, but there is still more to come immediately afterward.

**Comment by diehuxx** [comment]
cc: @angiejones

File: packages/http-server/src/in-memory-exchanges-api.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      for (const id of opts.filter.id) {
+        const exchange = this.exchangeMessagesMap.get(id)
+        if (exchange?.rfq?.from === opts.filter.from) {
+          exchanges.push(exchange)
+        }
+      }
+    } else {
+      // filter only has `from`
+      this.exchangeMessagesMap.forEach((exchange, _id) => {
+        // You definitely shouldn't use FakeExchangesApi in production.
```
Comment (KendallWeihe) [inline packages/http-server/src/in-memory-exchanges-api.ts:39]:
```suggestion
        // You definitely shouldn't use InMemoryExchangesApi in production.
```

File: packages/http-server/src/request-handlers/create-exchange.ts [hunk 4]
```diff
@@ -1,69 +1,70 @@
...
 
 type CreateExchangeOpts = {
-  callback: SubmitRfqCallback
+  callback?: SubmitRfqCallback
   offeringsApi: OfferingsApi
   exchangesApi: ExchangesApi
 }
 
-export function createExchange(options: CreateExchangeOpts): RequestHandler {
+export async function createExchange(req: Request, res: Response, options: CreateExchangeOpts): Promise<any> {
```
Comment (KendallWeihe) [inline packages/http-server/src/request-handlers/create-exchange.ts:14]: Alternative approach... 

```suggestion
export async function createExchange(req: Request, res: Response, options: CreateExchangeOpts): Promise<void> {
```

And then rather than something like this,

```typescript
return res.status(400).json({ errors: [{ detail: 'replyTo must be a valid url' }] })
```

We would do this...

```typescript
res.status(400).json({ errors: [{ detail: 'replyTo must be a valid url' }] })
return 
```

Benefit here is the function signature is intentional in how we expect it to be used (but maybe I'm wrong, maybe we want calling functions to have access to the `Response` type which is currently being returned?)

Comment (KendallWeihe) [reply packages/http-server/src/request-handlers/create-exchange.ts:14]: Oh shoot, disregard, this is intended to be used within express's dependencies

**Comment by KendallWeihe** [comment]
This all makes sense to me (I only briefly glossed over the tests)

> I'm deliberately not adding a changeset to this PR since I don't want to release a new version at this point. This PR adds a lot of tests and cleanup, but there is still more to come immediately afterward.

I think it may be good to list any breaking changes (changes to function signatures -- I call these "binary incompatibilities"), even if it's here in GitHub. But also, I'm not trying to promote unnecessary toil so I'm supportive of whatever we think is best, We have to manually cut releases so it's not like merging this will immediately break anything.

**Comment by mistermoe** [comment]
yooo @diehuxx 

> Refactor how TbdexHttpServer calls the callbacks because the previous way actually did not work.

what was the issue?

File: packages/http-server/tests/get-exchanges.spec.ts [hunk 5]
```diff
@@ -24,56 +26,137 @@ describe('GET /exchanges', () => {
...
+
+    const resp = await fetch(`http://localhost:8000/exchanges`, {
+      headers: {
+        'Authorization': `Bearer ${requestToken}`
+      }
+    })
+    expect(resp.status).to.equal(200)
+
+    expect(callbackSpy.callCount).to.eq(1)
+    // TODO: Check what arguments are passed to callback after we finalize its behavior
```
Comment (jiyoontbd) [inline packages/http-server/tests/get-exchanges.spec.ts:160]: do we need `callbackSpy.restore()` here or nah?

File: packages/http-server/tests/submit-close.spec.ts [hunk 6]
```diff
@@ -117,7 +145,220 @@ describe('POST /exchanges/:exchangeId/close', () => {
...
+    const resp = await fetch('http://localhost:8000/exchanges/123/close', {
+      method : 'POST',
+      body   : JSON.stringify(close)
+    })
+
+    expect(resp.status).to.equal(202)
+  })
+
+  it('returns a 202 if close is created by pfi', async () => {
+    // scenario: Alice creates an exchange and submits a Close message
```
Comment (jiyoontbd) [inline packages/http-server/tests/submit-close.spec.ts:222]:
```suggestion
    // scenario: Alice creates an exchange and PFI submits a Close message
```

**Review by jiyoontbd — APPROVED** [review]
nothing blocking, just a couple q's and comments! awesome work 🎉

**Review by mistermoe — APPROVED** [review]

**Comment by diehuxx** [comment]
> yooo @diehuxx
> 
> > Refactor how TbdexHttpServer calls the callbacks because the previous way actually did not work.
> 
> what was the issue?

TL;DR functional programming is confusing.

Previously, in the `listen` method, we set the routes for the server like so:
```typescript
this.api.get('/offerings', getOfferings({
      callback: this.callbacks['offerings'], offeringsApi
    }))
```
This calls `this.callbacks['offerings']` at the time when `server.listen()` is called, which means that `this.callbacks['offerings']` is always undefined because we haven't set our callbacks yet. What we actually want to do is get the callback when we handle an incoming message. I've done that like so:
```typescript
    this.api.get('/offerings', (req, res) =>
      getOfferings(req, res, {
        callback: this.callbacks['offerings'],
        offeringsApi
      })
    )
```
The arrow method `(req, res) => ...` isn't called until the server handles an incoming request, which means `this.callbacks['offerings']` isn't called until we handle a request. This allows us to do this:

```typescript
const api = new TbdexHttpServer()
const server = api.listen()
api.onSubmitRfq(() => console.log('We received an RFQ'))
// When an RFQ is submitted, it prints
// "We received an RFQ"

// We can also overwrite the old callback
api.onSubmitRfq(() => console.log('TBDex is my best friend'))
// When an RFQ is submitted, it prints
// "TBDex is my best friend"
```

File: packages/http-server/src/request-handlers/submit-close.ts [hunk 7]
```diff
@@ -1,55 +1,72 @@
...
+    return res.status(409).json({ errors: [errorResponse] })
+  }
+
+  // Ensure that Close is from either Alice or PFI
+  const rfq = exchange.rfq!
+  if (close.metadata.from === rfq.metadata.from && close.metadata.to === rfq.metadata.to) {
+    // Alice may Close an exchange
+  } else if (close.metadata.from === rfq.metadata.to && close.metadata.to === rfq.metadata.from) {
+    // The PFI may Close an exchange
+  } else {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/submit-close.ts:49]: This was a stylistic choice to write it this way because this felt the most readable and easy to document what's happening. I can refactor it to remove the empty if-self if you prefer -- the empty if-else is more of a personal preference on my part.

File: packages/http-server/src/request-handlers/get-offerings.ts [hunk 8]
```diff
@@ -1,23 +1,29 @@
...
 }
 
-export function getOfferings(opts: GetOfferingsOpts): RequestHandler {
+export async function getOfferings(request: Request, response: Response, opts: GetOfferingsOpts): Promise<any> {
   const { callback, offeringsApi } = opts
 
-  return async function (request, response) {
-    const queryParams = request.query as GetOfferingsFilter
-    const offerings = await offeringsApi.getOfferings({ filter: queryParams || {} })
+  const filter: GetOfferingsFilter = {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/get-offerings.ts:12]: Using `as` in typescript is like telling the compiler "just trust me bro", so it's good to avoid it unless we as programmers are pretty confident. Since `request.query` could realistically be anything, it's better to select the specific fields we want and use `filter: GetOfferingsFilter` to help us remember which fields we need.

File: packages/http-server/src/request-handlers/get-exchanges.ts [hunk 9]
```diff
@@ -1,49 +1,54 @@
...
+      queryParams.id = [request.query.id.toString()]
     }
+  }
 
-    return response.status(200).json({ data: exchanges })
+  const exchanges = await exchangesApi.getExchanges({ filter: queryParams })
+
+  if (callback) {
+    // TODO: figure out what to do with callback result. should we pass through the exchanges we've fetched
+    //       and allow the callback to modify what's returned? (issue #10)
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/get-exchanges.ts:49]: Haha yea I'm kinda rehashing old stuff, but I think we need open it back up for discussion. The PR that closed out #10 was #91, which did not include any documentation of rationale or tests, so I'm not convinced that we've thought through callback behavior in detail.

File: packages/http-server/src/request-handlers/get-exchanges.ts [hunk 10]
```diff
@@ -1,49 +1,54 @@
...
+  const queryParams: GetExchangesFilter = {
+    from: requesterDid,
+  }
 
-    if (callback) {
-      // TODO: figure out what to do with callback result. should we pass through the exchanges we've fetched
-      //       and allow the callback to modify what's returned? (issue #10)
-      const _result = await callback({ request, response }, queryParams)
+  if (request.query.id !== undefined) {
+    if (Array.isArray(request.query.id)) {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/get-exchanges.ts:38]: I'm open to the idea. Here's an issue in `tbdex` for it https://github.com/TBD54566975/tbdex/issues/239

File: packages/http-server/src/request-handlers/create-exchange.ts [hunk 11]
```diff
@@ -1,69 +1,70 @@
...
+import type { OfferingsApi, ExchangesApi, SubmitRfqCallback } from '../types.js'
 import { Rfq } from '@tbdex/protocol'
 import type { ErrorDetail } from '@tbdex/http-client'
 
 import { CallbackError } from '../callback-error.js'
+import { Request, Response } from 'express'
 
 type CreateExchangeOpts = {
-  callback: SubmitRfqCallback
+  callback?: SubmitRfqCallback
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/create-exchange.ts:9]: We were always treating it as nullable, but the unstrict typescript let us get away with incorrect typing

File: packages/http-server/src/request-handlers/create-exchange.ts (hunk 4 above)
Comment (diehuxx) [reply packages/http-server/src/request-handlers/create-exchange.ts:14]: I'm updating to `Promise<void>` since express.js doesn't look at any return type from this function. It seems to be a pure side effect. Express docs: https://expressjs.com/en/5x/api.html#app.get

File: packages/http-server/src/in-memory-exchanges-api.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    const exchange = this.exchangeMessagesMap.get(opts.exchangeId)
+    return exchange?.order
+  }
+
+  async getOrderStatuses(opts: { exchangeId: string }): Promise<OrderStatus[]> {
+    const exchange = this.exchangeMessagesMap.get(opts.exchangeId)
+    if (exchange?.orderstatus === undefined) {
+      return []
+    }
+    return [exchange.orderstatus]
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/in-memory-exchanges-api.ts:75]: Updating to have `Exchange` store an array of order statuses which will be returned from this function. Relatedly, I'm updating `OrderStatus.validNext` to contain `'orderstatus'`.

File: packages/http-server/tests/get-exchanges.spec.ts (hunk 5 above)
Comment (diehuxx) [reply packages/http-server/tests/get-exchanges.spec.ts:160]: No, the spy doesn't need restoring/resetting since we only use it in this one test and the spy is not replacing an existing function, it's spying on an anonymous function we created just for this test.

File: packages/http-server/tests/create-exchange.spec.ts [hunk 13]
```diff
@@ -1,18 +1,23 @@
...
 import { expect } from 'chai'
-
-let api = new TbdexHttpServer()
-let server: Server
+import { InMemoryExchangesApi } from '../src/in-memory-exchanges-api.js'
+import { InMemoryOfferingsApi } from '../src/in-memory-offerings-api.js'
+import { PortableDid } from '@web5/dids'
+import Sinon from 'sinon'
 
 describe('POST /exchanges/:exchangeId/rfq', () => {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/tests/create-exchange.spec.ts:11]: Which test(s) specifically are failing? I haven't seen any flakiness but I can investigate if you let me know where to look.

File: packages/http-server/src/request-handlers/create-exchange.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/request-handlers/create-exchange.ts:14]: I just pushed a commit changing the return type of this function to `Promise<void>`, explanation in [this comment thread](https://github.com/TBD54566975/tbdex-js/pull/170#discussion_r1493027472).

We're not longer using `RequestHandler` because the usage of this function has changed (see [this comment](https://github.com/TBD54566975/tbdex-js/pull/170#issuecomment-1946877790) for details). Previously, `createExchange` would return a function, which would be consumed by the express.js router like this:

```typescript
    this.api.post('/exchanges/:exchangeId/rfq', createExchange({
      callback: this.callbacks['rfq'], offeringsApi, exchangesApi,
    }))
```

Now, `createExchange` does NOT return a function. It is consumed as a curried function by the express js router like this
```typescript
    this.api.post('/exchanges/:exchangeId/rfq', (req: Request, res: Response) =>
      createExchange(req, res, {
        callback: this.callbacks['rfq'],
        offeringsApi,
        exchangesApi,
      })
```

File: packages/http-server/src/in-memory-exchanges-api.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    const exchange = this.exchangeMessagesMap.get(opts.exchangeId)
+    return Promise.resolve(exchange?.quote)
+  }
+
+  async getOrder(opts: { exchangeId: string }): Promise<Order | undefined> {
+    const exchange = this.exchangeMessagesMap.get(opts.exchangeId)
+    return exchange?.order
+  }
+
+  async getOrderStatuses(opts: { exchangeId: string }): Promise<OrderStatus[]> {
```
[the remark this answers is not in the document]
Comment (diehuxx) [reply packages/http-server/src/in-memory-exchanges-api.ts:70]: Good catch on the `validNext` thing. I was on the fence for this PR, but since you mention it, I'll add `orderstatus` as a valid next of `OrderStatus`. And I'm updating `getOrderStatuses` to return the array of order statuses.
